overlord/snapstate/backend/copydata.go

```diff
@@ -59,7 +59,27 @@ func (b Backend) UndoCopySnapData(newInfo *snap.Info, oldInfo *snap.Info, meter
 		if err2 != nil {
 			logger.Noticef("Cannot remove common data directories for %q: %v", newInfo.Name(), err2)
 		}
+	} else {
+		err2 = b.untrashData(newInfo)
+		if err2 != nil {
+			logger.Noticef("Cannot restore data from trash for %q: %v", newInfo.Name(), err2)
+		}
 	}
 
 	return firstErr(err1, err2)
 }
+
+// ClearTrashedData removes the trash. It returns no errors on the assumption that it is called very late in the game.
+func (b Backend) ClearTrashedData(oldSnap *snap.Info) {
+	dirs, err := snapDataDirs(oldSnap)
+	if err != nil {
+		logger.Noticef("Cannot clear trash for %q: %v", oldSnap.Name(), err)
+		return
+	}
+
+	for _, d := range dirs {
+		if err := clearTrash(d); err != nil && !os.IsNotExist(err) {
+			logger.Noticef("Problem clearing trash for %q's directory %s: %v", oldSnap.Name(), d, err)
```

Comment:
In this case I think we can just talk about the directory itself:

`("Cannot remove %s: %v", d, err)`